Module 18 badges

hocus2pocus opened the pull request:
Created badges

hocus2pocus committed:
Tests without questions are hidden from non-admins

hocus2pocus committed:
Created Badges Control Service

hocus2pocus committed:
Created view for badges

hocus2pocus committed:
Created button for badges index, minor style changes

hocus2pocus committed:
Created badges logic

BubuntuClu commented on app/services/badges_control_service.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def call
+    return unless @test_passage.success?
```
можно не делать две разные таблицы, бейджи и бейджи для пользователя. можно сделать связь многие-ко-многим. сделать еще поле `rule` или что-то подобное. 
тут у тебя тогда вызов можно будет заменить на
```
Badge.each do | badge|
   user.badges << badge if send "#{badge_rule}_award?", badge
end
```
имена методов делаешь под вызываемую конструкцию в `send`. в этом случае тебе не надо будет добавлять условия в код, если у тебя появится новый бейдж с новым методом, просто надо будет дописать метод обработчик и все заработает само.

BubuntuClu commented on app/services/badges_control_service.rb:
```diff
@@ -0,0 +1,89 @@
+    @test_passage = test_passage
+    @user = test_passage.user
+
+    badges_model_check
```
это можно вынести в админку, чтобы у админа была возможность создать новые бейджи

BubuntuClu replied:
это не задача сервиса создавать баджи. он проверяет выполнение условий. сами быджи ты можешь либо через админку создавать, либо вообще в сиде.

BubuntuClu commented on app/services/badges_control_service.rb:
```diff
@@ -0,0 +1,89 @@
+    all_tests_by_level - tests_from_test_passage == []
+  end
+
+  def achieve_first_attempt
```
эти методы не понядобиться, если сделать связь многие-ко-многим и использовать `user.badges << badge`

BubuntuClu commented on app/controllers/test_passages_controller.rb:
```diff
@@ -13,6 +13,7 @@ def update
 
     if @test_passage.completed?
       TestsMailer.completed_test(@test_passage).deliver_now
+      BadgesControlService.new(@test_passage).call
```
тут еще стоит проверку добавить, успешно ли тест пройден. если не успешно, то смысла выдавать бейдж нет.

BubuntuClu commented on app/services/badges_control_service.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def rule_all_by_category
+    # limit_date = previous_badge_date(@all_by_category) || @user.created_at
```
можно проще сделать через ассоциации

BubuntuClu commented on app/services/badges_control_service.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def rule_all_by_level
+    #current_budges = AchievedBadge.where(user_id: @user.id,)
```
`Test.where(level: level).count == @user.tests.where(level: level).uniq.count`

hocus2pocus committed:
Test

hocus2pocus committed:
Test

hocus2pocus committed:
Test

Merge branch 'Module_18_badges' of github.com:hocus2pocus/TestGuru into Module_18_badges

hocus2pocus committed:
Created BadgesUser join table

hocus2pocus committed:
Fixed BadgesControlService logic

hocus2pocus committed:
Creating default badges is moved to seed

hocus2pocus committed:
Fixed badges views

BubuntuClu commented on app/controllers/test_passages_controller.rb:
```diff
@@ -38,4 +39,24 @@ def gist
+
+    if badges.any?
+      current_user.badges << badges
+      badges.each do |badge|
```
тут тоже можно сделать без `case`, у тебя есть имя картинки, ты можешь его в символ конвертить `so_sym`.
с каждым новым баджем надо будет расширять этот кейс, этого в сервисе избежали, тут тоже надо

BubuntuClu commented on app/helpers/application_helper.rb:
```diff
@@ -12,6 +12,17 @@ def flash_helper(type)
   end
 
   def alert_type(alert_key)
-    alert_key == 'notice' ? 'success' : 'danger'
```
и здесь тоже бы от `case` избавиться

hocus2pocus committed:
Fixed badges index view
